File: asteroid.py
from numpy import random

class Asteroid:
  next_id = 0
  size_range = (500, 4000)
  distance_range = (10, 500)
  speed_range = (-10,10)

  def __init__(self, loc_constraint, initial_location = None, size = None, trajectory = None):
    self.id = self.__class__.next_id + 1
    self.__class__.next_id += 1
    self.initial_location = initial_location or (random.randint(loc_constraint[0]), random.randint(loc_constraint[1]), random.randint(*self.distance_range))
    self.size = size or random.randint(*self.size_range) #50, min([a / 4 for a in loc_constraint] + [500]))
    self.trajectory = trajectory or tuple(random.randint(*self.speed_range, size=3))
    self.hist_location = {0: self.initial_location}
    self.hist_radius = {0: self.size / 2}
    self.current_location = self.initial_location
    self.current_radius = self.size / 2

  def move(self, elapsed_seconds):
    self.current_location = self.location(elapsed_seconds)
    self.current_radius = self.radius(elapsed_seconds)

  def radius(self, elapsed_seconds):
    if not elapsed_seconds in self.hist_radius:
      self.hist_radius[elapsed_seconds] = self.size / self.location(elapsed_seconds)[2] / 2
    return self.hist_radius[elapsed_seconds]

  def location(self, elapsed_seconds):
    if not elapsed_seconds in self.hist_location:
      self.hist_location[elapsed_seconds] = \
        (self.initial_location[0] + self.trajectory[0] * elapsed_seconds,
         self.initial_location[1] + self.trajectory[1] * elapsed_seconds,
         self.initial_location[2] + self.trajectory[2] * elapsed_seconds)
    return self.hist_location[elapsed_seconds]

  def __repr__(self):
    return '<%s #%d initial_location: %s, size: %s, trajectory: %s >' % (
        self.__class__.__name__, self.id, self.initial_location, self.size,
        self.trajectory)

File: src/Asteroid.java
import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Random;

/**
 * Created by Sean on 2/4/17.
 */
public class Asteroid {
  private static final int[] size_range = {500, 4000};
  private static final int[] distance_range = {100, 1000};
  private static final int[] speed_range = {-5, 5};
  private static final String[] asteroid_images = {"asteroid_1.png", "asteroid_2.png", "asteroid_3.png"};

  private static int nextid = 0;

  private int[] initial_location;
  private int size;
  private int[] trajectory;
  private int id;
  private BufferedImage image;
  //private HashMap<Integer, Integer[]> hist_location;

  public int[] current_location;
  public int current_radius;

  public Asteroid(int[] initial_location, int size, int[] trajectory) {
    this.initial_location = initial_location;
    this.size = size;
    this.trajectory = trajectory;
    this.id = nextid;
    nextid += 1;
    setRandomImage();
  }

  public Asteroid(int[] loc_constraint) {
    Random rand = new Random();
    this.initial_location = new int[]{rand.nextInt(loc_constraint[0]), rand.nextInt(loc_constraint[1]), rand.nextInt(distance_range[1]) + distance_range[0]};
    this.size = rand.nextInt(size_range[1]) + size_range[0];
    this.trajectory = new int[]{rand.nextInt(speed_range[1]) + speed_range[0], rand.nextInt(speed_range[1]) + speed_range[0], rand.nextInt(speed_range[1]) + speed_range[0]};
    this.current_location = initial_location;
    this.current_radius = size / 2;
    this.id = nextid;
    nextid += 1;
    setRandomImage();
  }

  public int getId() {
    return this.id;
  }

  public void move(int elapsed_seconds) {
    this.current_location = location(elapsed_seconds);
    this.current_radius = radius(elapsed_seconds);
  }

  public int[] location(int elapsed_seconds) {
    return new int[]{initial_location[0] + trajectory[0] * elapsed_seconds,
                     initial_location[1] + trajectory[1] * elapsed_seconds,
                     initial_location[2] + trajectory[2] * elapsed_seconds};
  }

  public int radius(int elapsed_seconds) {
    if (location(elapsed_seconds)[2] == 0) {
      return 4000;
    }
    return (int)((size / 2.0) / (location(elapsed_seconds)[2] / 10.0));
  }

  public BufferedImage getImage() {
    return image;
  }

  @Override
  public String toString() {
    return "Asteroid{" +
        "id=" + id +
        ", initial_location=" + Arrays.toString(initial_location) +
        ", size=" + size +
        ", trajectory=" + Arrays.toString(trajectory) +
        ", current_location=" + Arrays.toString(current_location) +
        ", current_radius=" + current_radius +
        '}';
  }

  private void setRandomImage() {
    Random rand = new Random();
    String asteroid_image = asteroid_images[rand.nextInt(asteroid_images.length)];
    System.out.format("setting image to: %s\n", asteroid_image);
    try {
      this.image = ImageIO.read(new File("asteroids/" + asteroid_image));
    } catch (IOException e) {
    }
  }
}
